utoipa-gen/src/schema/component.rs

```diff
@@ -441,6 +482,85 @@ struct ComplexEnum<'a> {
     attributes: &'a [Attribute],
 }
 
+impl ComplexEnum<'_> {
+    fn unit_variant_tokens(variant_name: String) -> TokenStream2 {
+        quote! {
+            utoipa::openapi::PropertyBuilder::new()
+                .component_type(utoipa::openapi::ComponentType::String)
+                .enum_values::<[&str; 1], &str>(Some([#variant_name]))
+        }
+    }
+    /// Produce tokens that represent a variant of a [`ComplexEnum`].
+    fn variant_tokens(variant_name: String, variant: &Variant) -> TokenStream2 {
+        match &variant.fields {
+            Fields::Named(named_fields) => {
+                let named_enum = NamedStructComponent {
+                    attributes: &variant.attrs,
+                    fields: &named_fields.named,
+                    generics: None,
+                    alias: None,
+                };
+
+                quote! {
+                    utoipa::openapi::schema::ObjectBuilder::new()
+                        .property(#variant_name, #named_enum)
+                }
+            }
+            Fields::Unnamed(unnamed_fields) => {
+                let unnamed_enum = UnnamedStructComponent {
+                    attributes: &variant.attrs,
+                    fields: &unnamed_fields.unnamed,
+                };
+
+                quote! {
+                    utoipa::openapi::schema::ObjectBuilder::new()
+                        .property(#variant_name, #unnamed_enum)
+                }
+            }
+            Fields::Unit => Self::unit_variant_tokens(variant_name),
+        }
+    }
+
+    /// Produce tokens that represent a variant of a [`ComplexEnum`] where serde enum attribute
+    /// `tag = ` applies.
+    fn tagged_variant_tokens(tag: &str, variant_name: String, variant: &Variant) -> TokenStream2 {
+        match &variant.fields {
+            Fields::Named(named_fields) => {
+                let named_enum = NamedStructComponent {
+                    attributes: &variant.attrs,
+                    fields: &named_fields.named,
+                    generics: None,
+                    alias: None,
+                };
+
+                let variant_name_tokens = Self::unit_variant_tokens(variant_name);
+
+                quote! {
+                    #named_enum
+                        .property(#tag, #variant_name_tokens)
+                        .required(#tag)
+                }
+            }
+            Fields::Unnamed(_) => {
+                abort!(
+                    variant,
+                    "Unnamed (tuple) enum variants are unsupported for internally tagged enums using the `tag = ` serde attribute";
+
+                    help = "Try using a different serde enum representation";
+                );
```

Comment:
Awesome, But is't it possible to use unnamed fields with tag? Atleast with https://serde.rs/enum-representations.html#adjacently-tagged it is possible. So I guess it also should be possible with internally tagged as well? So this is maybe something that should be implemented though? 

Also now, if user has a type like one below, user will get the Error above and will not be able to create a complex enum when there is a `tag` defined. 
```rust
#[derive(Component)]
#[serde(tag = "my_tag")]
enum Foo {
   Bar,
   BarFoo {
      field: String,
   },
   Unnamed(String) // <--- causes the error above
}
```

The unnamed types ultimately will render as the inner type. So if it had a `tag = "my_tag"` I guess it should actually create object out of it similarly to here https://github.com/juhaku/utoipa/pull/146/files#diff-c2c506c1d61101323c309d3947b6328198fda433dd1be11232fab56d74345b02R515 But also adding the `.required(#tag)` to it as well as you have done already.

Reply:
Okay, then we can also forget them too for now. 

>We don't support adjacently tagged with the content attribute yet though?

Yes they are not implemented as  well, support for them could be done, and I think would be good idea for long run. Maybe this can be done in antoher PR so this wont grow too big.

Reply:
Great, yeah that's what I was thinking too :slightly_smiling_face: